### PR TITLE
[Storage] StateMerkleDB Lazy Commit

### PR DESCRIPTION
**Pull request by lightmark**
### Description
This would be the last PR of lazy commit feature. The current PR includes all the changes but no testing has been added. I'll add tests for async commit only and turn on sync_commit flag for most test cases to ensure they are not broken.

This PR does:
1. Put ledger_db and state_merkle_db into a struct called `StateDb` and  `impl DbReader for StateDb` instead of `StateStore` to make the next item easier. 
2. Refactored the state store initialization procedure to cater the change in 1. So now we don't have to create a dummy empty `BufferedState` just for a StateStore instance as a `DbReader` needed by `InMemoryStateCalculator` to run `write_set` at spinning up phase.
3. Changed `BufferedState` to include a pending commit `StateDelta` and an in-memory `StateDelta` that is not ready to commit because it doesn't cover any state checkpoint.
4. Created a async commit thread in background writing `StateDelta` onto disk.
5. Supported synchronously save_transactions by adding a flag that can blocks main thread until all the pending commits in background get written to disk.
6. Deleted the async commit thread in executor-benchmark.
7. Refactored storage-interface to remove `save_transactions_ext` method.
8. Changed `RocksdbConfigs` to `AptosDbConfig` and put `snapshot_size_threshold` into it as a config, leveraged by the unit tests.
9. Made `test_save_blocks` and `test_sync_transactions` in proptests test async feature with random-sized `snapshot_size_threshold`.
10. Remove `NoProofFetcher`, using `DbStateView` instead.

### Test Plan
changed two unit tests in proptests to test async commit has the expected behavior.

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2129)

### CODE REVIEW

File: storage/aptosdb/src/state_merkle_db.rs [hunk 1]
```diff
@@ -92,6 +94,54 @@ impl StateMerkleDb {
...
+        let mut batch = SchemaBatch::new();
+        {
+            let _timer = OTHER_TIMERS_SECONDS
+                .with_label_values(&["serialize_jmt_commit"])
+                .start_timer();
+
+            add_node_batch(
+                &mut batch,
+                tree_update_batch
+                    .node_batch
```
Comment (grao1991) [inline storage/aptosdb/src/state_merkle_db.rs:123]: also include the changes in https://github.com/aptos-labs/aptos-core/pull/2115?

File: storage/aptosdb/src/state_merkle_db.rs [hunk 2]
```diff
@@ -92,6 +94,54 @@ impl StateMerkleDb {
...
+                .flatten()
+                .map(|row| batch.put::<StaleNodeIndexSchema>(row, &()))
+                .collect::<Result<Vec<()>>>()?;
+        }
+
+        // commit jellyfish merkle nodes
+        let _timer = OTHER_TIMERS_SECONDS
+            .with_label_values(&["commit_jellyfish_merkle_nodes"])
+            .start_timer();
+        self.write_schemas(batch)?;
```
Comment (grao1991) [inline storage/aptosdb/src/state_merkle_db.rs:141]: move db commit to another async thread?

Comment (lightmark) [reply storage/aptosdb/src/state_merkle_db.rs:146]: another PR

File: storage/aptosdb/src/state_store/mod.rs [hunk 3]
```diff
@@ -48,7 +48,10 @@ mod state_store_test;
 type StateValueBatch = aptos_jellyfish_merkle::StateValueBatch<StateKey, StateValue>;
 
 pub const MAX_VALUES_TO_FETCH_FOR_KEY_PREFIX: usize = 10_000;
-const MAX_WRITE_SETS_AFTER_SNAPSHOT: LeafCount = 200_000;
+// We assume TARGET_SNAPSHOT_INTERVAL_IN_VERSION > block size.
+const MAX_WRITE_SETS_AFTER_SNAPSHOT: LeafCount = buffered_state::TARGET_SNAPSHOT_INTERVAL_IN_VERSION
```
Comment (msmouse) [inline storage/aptosdb/src/state_store/mod.rs:52]: 👍

**Review by msmouse — APPROVED** [review]

**Pull request merged**